pgjdbc/src/main/java/org/postgresql/core/Tuple.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2017, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+/**
+ * Class representing a row in a {@link java.sql.ResultSet}.
+ */
+public class Tuple {
+  private final boolean forUpdate;
+  final byte[][] data;
+
+  /**
+   * Construct an empty tuple. Used in updatable result sets.
+   * @param length the number of fields in the tuple.
+   */
+  public Tuple(int length) {
+    this(new byte[length][], true);
+  }
+
+  /**
+   * Construct a populated tuple. Used when returning results.
+   * @param data the tuple data
+   */
+  public Tuple(byte[][] data) {
+    this(data, false);
+  }
+
+  private Tuple(byte[][] data, boolean forUpdate) {
+    this.data = data;
+    this.forUpdate = forUpdate;
+  }
+
+  /**
+   * Number of fields in the tuple
+   * @return number of fields
+   */
+  public int fieldCount() {
+    return data.length;
+  }
+
+  /**
+   * Total length in bytes of the tuple data.
+   * @return the number of bytes in this tuple
+   */
+  public int length() {
```

Comment:
should we cache this?

Reply:
Don't see any reason not to.

Reply:
This will assume that we're not mutating the incoming `byte[][]` from the outside - which would probably be problematic in other areas as well

Reply:
Currently it seems to be only used only here: https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java#L2225-L2239

Is it still worth caching it?

Reply:
If it's only ever called once then there is no reason to cache it.